lib/aws-s3-stack.ts

```diff
@@ -1,8 +1,56 @@
 import * as cdk from 'aws-cdk-lib';
 import { Construct } from 'constructs';
+import { AwsS3StackProps } from './AwsS3StackProps';
+import * as s3 from 'aws-cdk-lib/aws-s3';
 
 export class AwsS3Stack extends cdk.Stack {
-  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
+  /**
+   * Constructs a new instance of the AwsS3Stack.
+   *
+   * @param scope The parent construct.
+   * @param id An identifier for the stack.
+   * @param props Configuration properties for the stack, including resource prefixes, bucket names, and deployment environment.
+   */
+  constructor(scope: Construct, id: string, props: AwsS3StackProps) {
     super(scope, id, props);
+
+    const removalPolicy = props.deployEnvironment === 'production' ? cdk.RemovalPolicy.RETAIN : cdk.RemovalPolicy.DESTROY;
+
+    // define an S3 bucket
+    const s3Bucket = new s3.Bucket(this, `${props.resourcePrefix}-${props.s3BucketName}`, {
+      bucketName: `${props.deployEnvironment}-${props.deployRegion}-${props.s3BucketName}`,
+      encryption: s3.BucketEncryption.S3_MANAGED,
+      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
+      publicReadAccess: false,
+      removalPolicy: removalPolicy,
+      autoDeleteObjects: removalPolicy === cdk.RemovalPolicy.DESTROY,
+      accessControl: s3.BucketAccessControl.BUCKET_OWNER_FULL_CONTROL,
+      versioned: true, // Enable versioning
+      serverAccessLogsBucket: new s3.Bucket(this, `${props.resourcePrefix}-s3-bucket-logs`, {
```

Comment:
**Suggestion:** Consider implementing lifecycle rules for the S3 bucket to manage object transitions and expirations. This can help optimize storage costs by moving older objects to cheaper storage tiers or deleting them after a certain period.
> 🌕  **High Priority** 
 >[Category: **`Best Practice`**, Importance: **`9`**]
> **Reason:** This suggestion introduces lifecycle rules for the S3 bucket, which is a best practice for optimizing storage costs and managing object expiration. It is highly relevant and improves the functionality of the S3 bucket configuration without introducing any breaking changes.

```suggestion
    const s3Bucket = new s3.Bucket(this, `${props.resourcePrefix}-${props.s3BucketName}`, {
      bucketName: `${props.deployEnvironment}-${props.deployRegion}-${props.s3BucketName}`,
      encryption: s3.BucketEncryption.S3_MANAGED,
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      publicReadAccess: false,
      removalPolicy: removalPolicy,
      autoDeleteObjects: removalPolicy === cdk.RemovalPolicy.DESTROY,
      accessControl: s3.BucketAccessControl.BUCKET_OWNER_FULL_CONTROL,
      versioned: true,
      lifecycleRules: [
        {
          transitions: [
            {
              storageClass: s3.StorageClass.INFREQUENT_ACCESS,
              transitionAfter: cdk.Duration.days(90),
            },
            {
              storageClass: s3.StorageClass.GLACIER,
              transitionAfter: cdk.Duration.days(180),
            },
          ],
          expiration: cdk.Duration.days(365),
        },
      ],
```
- [ ]  **I have reviewed this code suggestion and addressed the relevant ones.**